### ISSUE TITLE
How to actually use reselect's libdef?

### ISSUE THREAD

**Issue by callumlocke** [flowtype/flow-typed#2445]
The [reselect libdef](https://github.com/flowtype/flow-typed/blob/5d8678f464a6c28fe9f67738e2ecde1163180ff1/definitions/npm/reselect_v3.x.x/flow_v0.47.x-/reselect_v3.x.x.js) is complex and I'm not sure how to read it.

Does it give you some way to pass in generics when use use `createSelector`, e.g. the type of your state object that will be passed into the returned function, so Flow can figure out the types of arguments etc? How exactly should I do this? I've tried this but it causes hundreds of confusing errors:

```js
createSelector<State>(selector1, selector2, (one, two) => {...})
```

More generally, if a libdef does provide types that accept generics, and is complex enough to need to be documented, where should this documentation live?

**Comment by AndrewSouthpaw** [comment]
I'd suggest you tag people who have worked on that libdef, so they can answer your question.

**Comment by MatrixFrog** [comment]
@NilSet @jedwards1211 @julienw

**Comment by jedwards1211** [comment]
@callumlocke Flow automagically infers type parameters -- IIRC the syntax doesn't even support explicitly specifying the parameters in a function call like that (unlike generics in other languages like Java).  So just do `createSelector(selector1, selector2, (one, two) => {...})`, and if you need to, annotate the type of `selector1`, `selector2`, `one`, `two`, etc.  Any type errors that occur can be confusing...it's just kind of part of the territory, and there's not really anything the typedefs can do to make this better.

The reason the typedefs seem complex is just because they have to make a separate def for each number of arguments:
* `(selector1, (one) => {...})`
* `(selector1, selector2, (one, two) => {...})`
* `(selector1, selector2, selector3, (one, two, three) => {...})`
Because unfortunately there's currently no way in flow to make a single type def that can map the return types of the selector arguments `0..n-1` into arguments for the final function.

**Comment by pascalduez** [comment]
@jedwards1211 
> IRC the syntax doesn't even support explicitly specifying the parameters in a function call like that (unlike generics in other languages like Java). 

FWIW it's supported from version `0.72` if I'm not mistaken.    
https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVATApgYxgQwCcswoBXAOxwBcBLOC0uOAHgBUA+ACgA8AuMGwCUAtgG50UZiwDO1QrQoBzbgEYATAGYhYoA

**Comment by amccloud** [comment]
@jedwards1211 with flow@0.85 flow cannot infer like you mentioned and requires the new explicit syntax `createSelector<State, *, *, *>` for any selector that is exported

The odd part now is that it requires explicit definitions for all placeholders which is making the current libdefs almost impossible to use with the latest flow

```js
type State = {
  x: number
};

export const testSelector = createSelector(
  (state: State) => state.x,
  (x) => x + 1
)({ x: 0 });

export const testBrokenSelector = createSelector( // $FlowError Missing type annotation for TProps, TResult, TState ... 
  (state: State) => state.x,
  (x) => x + 1
);

export const testExplicitSelector = createSelector<State, *, *, *>(
  (state: State) => state.x,
  (x) => x + 1
);

export const testBrokenExplicitSelector = createSelector<State, *, *, *>(
  [
    (state: State) => state.x,
    (state: State) => state.y
  ], // $FlowError: Cannot call createSelector because array literal [1] has an arity of 2 but tuple type [2] has an arity of 1
  (x, y) => x + y
);
```

<details>
  <summary>Click to see full flow error</summary>
```sh
yarn run v1.12.3
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ test_reselect_3.x.x.js:18:35

Missing type annotation for TProps. TProps is a type parameter declared in function type [1] and was implicitly
instantiated at call of createSelector [2].

     test_reselect_3.x.x.js
     15│   (x) => x + 1
     16│ )({ x: 0 });
     17│
 [2] 18│ export const testBrokenSelector = createSelector( // $FlowError Missing type annotation for TProps, TResult, TState
     19│   (state: State) => state.x,
     20│   (x) => x + 1
     21│ );
     22│
     23│ export const testExplicitSelector = createSelector<State, *, *, *>(
     24│   (state: State) => state.x,

     flow-typed/npm/reselect_v3.x.x.js
 [1] 19│     <TState, TProps, TResult, T1>(
     20│       selector1: InputSelector<TState, TProps, T1>,
     21│       resultFunc: (arg1: T1) => TResult
     22│     ): OutputSelector<TState, TProps, TResult>,


Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ test_reselect_3.x.x.js:18:35

Missing type annotation for TResult. TResult is a type parameter declared in function type [1] and was implicitly
instantiated at call of createSelector [2].

     test_reselect_3.x.x.js
     15│   (x) => x + 1
     16│ )({ x: 0 });
     17│
 [2] 18│ export const testBrokenSelector = createSelector( // $FlowError Missing type annotation for TProps, TResult, TState
     19│   (state: State) => state.x,
     20│   (x) => x + 1
     21│ );
     22│
     23│ export const testExplicitSelector = createSelector<State, *, *, *>(
     24│   (state: State) => state.x,

     flow-typed/npm/reselect_v3.x.x.js
 [1] 19│     <TState, TProps, TResult, T1>(
     20│       selector1: InputSelector<TState, TProps, T1>,
     21│       resultFunc: (arg1: T1) => TResult
     22│     ): OutputSelector<TState, TProps, TResult>,


Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ test_reselect_3.x.x.js:18:35

Missing type annotation for TState. TState is a type parameter declared in function type [1] and was implicitly
instantiated at call of createSelector [2].

     test_reselect_3.x.x.js
     15│   (x) => x + 1
     16│ )({ x: 0 });
     17│
 [2] 18│ export const testBrokenSelector = createSelector(
     19│   (state: State) => state.x,
     20│   (x) => x + 1
     21│ );
     22│
     23│ export const testExplicitSelector = createSelector<State, *, *, *>(
     24│   (state: State) => state.x,

     flow-typed/npm/reselect_v3.x.x.js
 [1] 19│     <TState, TProps, TResult, T1>(
     20│       selector1: InputSelector<TState, TProps, T1>,
     21│       resultFunc: (arg1: T1) => TResult
     22│     ): OutputSelector<TState, TProps, TResult>,


Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ test_reselect_3.x.x.js:29:3

Cannot call createSelector because array literal [1] has an arity of 2 but tuple type [2] has an arity of 1.

     test_reselect_3.x.x.js
     25│   (x) => x + 1
     26│ );
     27│
     28│ export const testBrokenExplicitSelector = createSelector<State, *, *, *>(
 [1] 29│   [
     30│     (state: State) => state.x,
     31│     (state: State) => state.y
     32│   ],
     33│   (x, y) => x + y
     34│ );
     35│

     flow-typed/npm/reselect_v3.x.x.js
 [2] 24│       selectors: [InputSelector<TState, TProps, T1>],


Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ test_reselect_3.x.x.js:31:29

Cannot get state.y because property y is missing in State [1].

     28│ export const testBrokenExplicitSelector = createSelector<State, *, *, *>(
     29│   [
     30│     (state: State) => state.x,
 [1] 31│     (state: State) => state.y
     32│   ],
     33│   (x, y) => x + y
     34│ );



Found 5 errors
```
</details>

**Comment by jedwards1211** [comment]
You know, I just realized,  regardless of the new errors in Flow 0.85, the typedefs are totally wrong.  Selectors support any number of arguments, not just `state` and `props` (and they don't treat those two arguments any differently from the 3rd, 4th, etc arguments)

**Comment by jedwards1211** [comment]
@pascalduez good point, it's also important to note that Babel 6 doesn't support explicit type parameters on function invocations, for people who are stuck on Babel 6 for one reason or another.

**Comment by langri-sha** [comment]
Is it wrong if we use `createSelector<State, *, *, *>` when exporting selectors to work around Flow errors?

**Comment by yfr** [comment]
Is there any news on this? Can we update the libdefs? Or is there another way around. With all the possible args I sometimes need to write code like this: 

`createSelector<StoreState, *, *, *, *, *, *, *, *>`

Which looks rather funny. And I guess this is not what the inventor had in mind.

I would love to help but, although we are using them a lot and they help us a lot, types are still a very weird thing to think about for me.

**Comment by langri-sha** [comment]
Could it be possible to somehow reshape the types to trim off some of the parameterized types, depending on application? 

I can understand that someone might have a large Flow project with many different stores and selectors that sum over many input selectors that accept different props types, but I think there's a majority of us that have a single state and most of our selectors don't accept props. Does anyone have suggestions on how to improve this sort of an experience?

**Comment by julienw** [comment]
I haven't tried, but wouldn't the following work:

Instead of 
```js
export const testExplicitSelector = createSelector<State, *, *, *>(
  (state: State) => state.x,
  (x) => x + 1
);
```
do:
```js
export const testExplicitSelector: State => number = createSelector(
  (state: State) => state.x,
  (x) => x + 1
);
```

Again I haven't tried (in our project we still use a quite ancient version of flow), but this looks logical.

**Comment by DewarM** [comment]
Hi 👋 I am having the exact same issue as seen above:
```js
export const testBrokenExplicitSelector = createSelector<State, *, *, *>(
  [
    (state: State) => state.x,
    (state: State) => state.y
  ], // $FlowError: Cannot call createSelector because array literal [1] has an arity of 2 but tuple type [2] has an arity of 1
  (x, y) => x + y
);
```
I really don't understand, the typings are of a tuple type which, if I understand correctly: they are not compatible with arrays? https://flow.org/en/docs/types/tuples/#toc-tuples-don-t-match-array-types

If I change the type to 
```js
selectors: Array<InputSelector<TState, TProps, T1>>,
```
from 
```js
selectors: [InputSelector<TState, TProps, T1>],
```

I no longer get the error. But this causes `y` in the example above to have the `void` type...

How are folks dealing with this atm?

**Comment by jedwards1211** [comment]
@DewarM by sticking with Flow 0.84 😂

**Comment by DewarM** [comment]
@jedwards1211 I'll downgrade and see how that plays out cheers!

**Comment by artemlitch** [comment]
Has anyone gotten this to work? Looking to upgrade our flow for a variety of good reasons, but reselect is causing the same issues

**Comment by julienw** [comment]
This is weird, we have no problem at all with latest flow. We do type our selector functions with this type:
https://github.com/firefox-devtools/profiler/blob/ef2ba4da9e63653a8686e9d63957a2353f659d51/src/types/store.js#L29
Maybe this makes it work magically ?

**Comment by jedwards1211** [comment]
@julienw `type Selector<T> = State => T` would do absolutely nothing to ensure type correctness of everything in a reselect selector, it's not really relevant to this issue

**Comment by jgreen210** [comment]
I've got the reselect_v3.x.x.js libdef working in moderately large code base with flow 0.92. Had to:

* Specify createSelector() return types explicitly in many cases: https://github.com/flow-typed/flow-typed/issues/2445#issuecomment-465107183. Just using types like OurStateType => boolean in almost all cases. There was one place where had to use libdef's OutputSelector type. Not sure why, maybe since had a function that returned a family of selectors, it wasn't since used any of the extra methods defined by this type.
* Also had one createSelector() call that had too many arguments for the libdef (more than 16) where I had to suppress the flow warning. Didn't feel like rewriting the code to use fewer arguments or modifying the libdef.

I don't think we do anything particularly exotic with reselect in our code base, so there could easily be problems for other people.

The libdef picked up some useful typing errors in our code, so is worth using.

**Comment by jamesisaac** [comment]
As has been mentioned above, what's imported is annotating the exported signature, which means you just have to annotate the resulting selector itself (no need to worry about the internal `createSelect` call, that can be inferred).  So @jedwards1211 your example just needs one extra annotation, right at the module boundary:

```js
export const selectStuff: State => string = createSelector(
  ({ foo }: State) => foo,
  ({ bar }: State) => bar,
  (foo: number, bar: string): string => repeat(bar, foo)
)
```

Are there any examples that wouldn't be fairly simple to annotate in the above way?

You can see some recent discussion about Flow's decision here: https://github.com/facebook/flow/issues/7647#issuecomment-485527528 and it seems the Flow team are pretty committed to this direction.

**Comment by julienw** [comment]
> @julienw `type Selector<T> = State => T` would do absolutely nothing to ensure type correctness of everything in a reselect selector, it's not really relevant to this issue

As others have said, it seems to be enough to type the exported selector functions, and using such a type makes this slightly more convenient.

**Comment by IanVS** [comment]
It seems that if you specify the wrong type on the exported selector, flow will still throw an error.  It knows what `createSelector` is coming up with, it's now just forcing us to manually specify the result as well?

**Comment by julienw** [comment]
@ianvs Yeah, [Flow forces you to explicitely type all exported values](https://medium.com/flow-type/asking-for-required-annotations-64d4f9c1edf8). (read this post for a more detailed write-up that I could do :) ).
If you don't export a selector but merely use it in the same file, then you probably won't need to type it explicitely -- but in our project we found that it still helped to reason and that flow errors were easier to read and follow if we type all selectors explicitely.

**Comment by kangax** [comment]
Current reselect definitions are definitely wrong, from what I can tell. Replacing `[Selector<TState, TProps, T1>]` with `Array<Selector<TState, TProps, T1>>` just reduced the number of errors in our codebase from 1140 to 407. The second argument is not an array with 1 selector (as current definitions specify it) but an array of selectors.

**Comment by langri-sha** [comment]
Yes, can confirm something isn't right. I'm getting warnings of unused
suppression comments.

**Comment by Yohandah** [comment]
I'm doing like this to type my selector 
```js
export const selectStuff: State => string = createSelector(
  ({ foo }: State) => foo,
  ({ bar }: State) => bar,
  (foo: number, bar: string): string => repeat(bar, foo)
)
```

but this prevents me from using stuff like `resultFunc` on my selector ... 

<img width="712" alt="image" src="https://user-images.githubusercontent.com/15921402/223491868-b1e4a534-9765-412b-8c12-3e81e280d3da.png">

has anyone found a decent workaround ?
